internal/pkg/uploader/finalize.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	ErrFailValidation  = errors.New("file contents failed validation")
+	ErrStatusNoUploads = errors.New("file closed, not accepting uploads")
+)
+
+func (u *Uploader) Complete(ctx context.Context, id string, transitHash string) (upload.Info, error) {
+	// make sure document is freshly fetched, not cached
+	// so accurate status checking happens
+	info, err := FetchUploadInfo(ctx, u.bulker, id)
+	if err != nil {
+		return info, err
+	}
+
+	/*
+		Verify Upload
+	*/
+
+	// if already done, failed or deleted, exit
+	if !info.StatusCanUpload() {
+		return info, ErrStatusNoUploads
+	}
+
+	chunks, err := GetChunkInfos(ctx, u.bulker, info.DocID)
+	if err != nil {
+		return info, err
+	}
+	if !u.allChunksPresent(info, chunks) {
+		return info, ErrMissingChunks
+	}
+	if !u.verifyChunkInfo(info, chunks, transitHash) {
+		if err := SetStatus(ctx, u.bulker, info, upload.StatusFail); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error setting the upload status to failure")
+		}
+		if err := DeleteChunksByQuery(ctx, u.bulker, info.Source, info.DocID); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error deleting left-behind chunk data")
+		}
+		return info, ErrFailValidation
+	}
+
+	/*
+		Upload OK. Update status and save valid transithash
+	*/
+	if err := MarkComplete(ctx, u.bulker, info, transitHash); err != nil {
+		return info, err
+
+	}
+
+	return info, nil
+}
+
+func (u *Uploader) allChunksPresent(info upload.Info, chunks []ChunkInfo) bool {
+	// check overall count
+	if len(chunks) != info.Count {
+		log.Warn().Int("expectedCount", info.Count).Int("received", len(chunks)).Interface("chunks", chunks).Msg("mismatch number of chunks")
+		return false
+	}
+
+	// now ensure all positions are accounted for, no gaps, etc
+	sort.Slice(chunks, func(i, j int) bool {
+		return chunks[i].Pos < chunks[j].Pos
+	})
+
+	for i, c := range chunks {
+		if c.Pos != i {
+			log.Warn().Int("expected", i).Interface("chunk", c).Msg("chunk position doesn't match. May be a gap in uploaded file")
+			return false
+		}
+	}
+
+	return true
+}
+
+func (u *Uploader) verifyChunkInfo(info upload.Info, chunks []ChunkInfo, transitHash string) bool {
+	// verify all chunks except last are info.ChunkSize size
+	// verify last: false (or field excluded) for all except final chunk
+	// verify final chunk is last: true
+	// verify hash
+
+	hasher := sha256.New()
+
+	for i, chunk := range chunks {
+		if i < info.Count-1 {
+			// all chunks except last must have last:false
+			// and be PRECISELY info.ChunkSize bytes long
+			if chunk.Last {
+				log.Debug().Int("chunkID", i).Msg("non-final chunk was incorrectly marked last")
+				return false
+			}
+			if chunk.Size != int(info.ChunkSize) {
```

Comment:
should we just have the chunk size defined as int64(or uint64) for both structs and avoid casting? the int is int64 anyways, the fleet server is built for 64bit only atm